File: src/main/java/cc/mirukuneko/realtrainmodrenewed/network/SpeakerStopPayload.java
package cc.mirukuneko.realtrainmodrenewed.network;

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed;
import cc.mirukuneko.realtrainmodrenewed.client.sound.LegacyScriptSoundManager;
import io.netty.buffer.ByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import net.neoforged.neoforge.network.handling.IPayloadContext;

/**
 * スピーカーブロックが壊された等で、指定座標で再生中の音を止めるパケット。
 * クライアントは {@link LegacyScriptSoundManager#stopAt} でその位置の音を停止する
 * (長い音がブロック破壊後も鳴り続ける問題の対策)。
 */
public record SpeakerStopPayload(double x, double y, double z) implements CustomPacketPayload {

    public static final Type<SpeakerStopPayload> TYPE = new Type<>(
        Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "speaker_stop")
    );

    public static final StreamCodec<ByteBuf, SpeakerStopPayload> STREAM_CODEC = StreamCodec.composite(
        ByteBufCodecs.DOUBLE, SpeakerStopPayload::x,
        ByteBufCodecs.DOUBLE, SpeakerStopPayload::y,
        ByteBufCodecs.DOUBLE, SpeakerStopPayload::z,
        SpeakerStopPayload::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    public static void handleOnClient(SpeakerStopPayload payload, IPayloadContext context) {
        context.enqueueWork(() ->
            LegacyScriptSoundManager.stopAt(payload.x(), payload.y(), payload.z())
        );
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/network/SpeakerStopPayload.kt
package cc.mirukuneko.realtrainmodrenewed.network

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.client.sound.LegacyScriptSoundManager
import io.netty.buffer.ByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import net.neoforged.neoforge.network.handling.IPayloadContext

data class SpeakerStopPayload(
    val x: Double,
    val y: Double,
    val z: Double,
) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<SpeakerStopPayload> = CustomPacketPayload.Type(
            Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "speaker_stop")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, SpeakerStopPayload> = StreamCodec.composite(
            ByteBufCodecs.DOUBLE,
            { payload -> payload.x },
            ByteBufCodecs.DOUBLE,
            { payload -> payload.y },
            ByteBufCodecs.DOUBLE,
            { payload -> payload.z },
            ::SpeakerStopPayload,
        )

        @JvmStatic
        fun handleOnClient(payload: SpeakerStopPayload, context: IPayloadContext) {
            context.enqueueWork {
                LegacyScriptSoundManager.stopAt(payload.x, payload.y, payload.z)
            }
        }
    }
}
